File: leetcode/377.combination-sum-iv/377.combination-sum-iv.py
#
# @lc app=leetcode id=377 lang=python3
#
# [377] Combination Sum IV
#

# @lc code=start
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[0] = 1
        for i in range(target + 1):
            for x in nums:
                if i >= x:
                    dp[i] = dp[i] + dp[i - x]
        return dp[target] 
# @lc code=end

File: leetcode/377.combination-sum-iv/377.combination-sum-iv.cpp
/*
 * @lc app=leetcode id=377 lang=cpp
 *
 * [377] Combination Sum IV
 */

// @lc code=start
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        unordered_set<int> s;
        vector<unsigned int> dp(target + 1, 0);
        dp[0] = 1;
        for (auto num : nums) {
            s.insert(num);
        }

        for (int i = 1; i < dp.size(); ++i) {
            for (int j = i - 1; j >= 0; --j) {
                if (s.find(i - j) != s.end()) {
                    dp[i] += dp[j];
                }
            }
        }

        return dp[target];
    }
};
// @lc code=end

